### PR TITLE
[feat] application versioning and update detection support

### PR DESCRIPTION
**Pull request by pz-mxu**
This PR will offer two options for navigation errors in the router. These errors occur for example if a new version of the site is deployed while being open in a browser. The JS files for pages might have changed and so has the filename. The previous page JS files, that the router still knows about, will give an import error 500 when trying to navigate to that page. For more detail on the problem, see #87.

The `kit.browser.router` configuration option in svelte.conf.js is now an object with `enabled` and `onError` properties.

`enabled` is just a boolean that replaces the old `kit.browser.router` boolean option.

`onError` has two possible values

- "fail": will result in exactly the same behaviour as now - import error 500
- "reload": location.href will be set to force a reload when navigation failed and the app version changed

The "reload" option should be enough for most users with the problem described. "fail" is currently the default option.

The mentioned app version is an automatically generated timestamp, that can be accessed with 

`import { version} from "$app/env"`

and can also be overridden by setting the environment variable `VITE_APP_VERSION` to a custom version string.

There is also an `updated` store that can be imported with

`import { updated } from "$app/stores"`.

The app version will be checked periodically in the background by fetching "_app/version.json" every 5 minutes. The store is set to `true` when the fetched version differs from the running version. The store has also a `check` function, that will immediately perform the check and return the result as a promise.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by benmccann** [comment]
> "redirect": location.href will be set to force a reload when navigation failed

I didn't look at the code, but just from the description, it sounds like "reload" would be a better name. That being said, I think that would rarely be the best behavior and not something I'd want to encourage, so I'd be much happier to remove this option and have the user implement it themselves (e.g. via the `function`)

**Comment by pz-mxu** [comment]
> That being said, I think that would rarely be the best behavior and not something I'd want to encourage, so I'd be much happier to remove this option and have the user implement it themselves (e.g. via the `function`)

The implementation is "reload" is trivial, so that's fine for me.

**Comment by Rich-Harris** [comment]
Thank you! Personally I disagree about reloading not being the best option — I think it's _exactly_ what you want to happen if you attempt to click a link (or `goto`) and the module can't be imported (unless it's because you're offline, in which case rendering the error page is appropriate). I'm not sure what other behaviour you'd want to implement?

The trouble, of course, is that the error might be happening for some other reason than a new deploy — perhaps the imported module immediately throws an error in the browser, or something. If we reload indiscriminately then we miss the opportunity to discover those cases. Instead, we need a cohesive idea of an app 'version' (that can be used to disable client-side routing pre-emptively, or show a toast, or assess whether API requests are up to date, etc) that is used to determine whether reloading is appropriate.

Stringifying functions is a no-go, by the way — it gets horribly confusing when suddenly closures don't work, etc. If we need that level of flexibility (I'm not sure we do, yet) then we need client-side hooks.

**Comment by pz-mxu** [comment]
> ...we need a cohesive idea of an app 'version' (that can be used to disable client-side routing pre-emptively, or show a toast, or assess whether API requests are up to date, etc) that is used to determine whether reloading is appropriate.
I added an app version. The reload will only happen now on error when the version changed.

> Stringifying functions is a no-go, by the way — it gets horribly confusing when suddenly closures don't work, etc. If we need that level of flexibility (I'm not sure we do, yet) then we need client-side hooks.
Understood, so for now I removed the custom function option.

File: packages/kit/src/core/build/build_client.js
Comment (Rich-Harris) [inline packages/kit/src/core/build/build_client.js:32]: We need a way to specify the version string via config. It should fall back to a timestamp (`Date.now()` is fine, I think — no need for an ISO string) but I've seen cases where the same app version is built multiple times, and the builds disagree. In those cases you need to be able to specify a commit hash or a `package.json` version or whatever.

File: packages/kit/src/core/build/build_client.js
Comment (Rich-Harris) [inline packages/kit/src/core/build/build_client.js:112]: To make this future-proof it should be an object:

```suggestion
		JSON.stringify({ version: process.env.VITE_APP_VERSION })
```

File: packages/kit/src/core/build/build_client.js
Comment (Rich-Harris) [inline packages/kit/src/core/build/build_client.js:113]: No need to specify encoding with `writeFileSync`
```suggestion
```

File: packages/kit/src/runtime/client/renderer.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/renderer.js:159]: This will need to be configurable

File: packages/kit/types/internal.d.ts
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:10]: guessing this is left over from something?

File: packages/kit/src/runtime/client/renderer.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/renderer.js:74]: This needs to use `config.kit.appDir` somehow, it can't be hardcoded to `_app`.

Also, no need to create a `Headers` object, you can just do this:

```js
const new_version - await fetch(`${base}/${app_dir}/version.json`, {
  headers: {
    'pragma': 'no-cache',
    'cache-control': 'no-cache'
  }
})
```

We shouldn't be swallowing the error here — at the very least an explicit call to `await updated.check()` should throw if the fetch fails.

**Review by Rich-Harris — CHANGES_REQUESTED** [review]
Thank you. Thinking on this further, I'm reaching the conclusion that there's no reason for any behaviour other than automatic reloading. This is an area where the framework needs to be a little bit opinionated and do the right thing, since most developers don't really think about long session handling until they get bitten. Since client-side routing is best thought of as a progressive enhancement, falling back to a full page refresh should be perfectly acceptable in a well-written app.

To that end, I think we should get rid of the `onError` option (though if there's a compelling reason not to, I would be interested to hear it).

As mentioned inline, we need to be able to configure both the version string and the polling interval. I propose the following:

```js
// svelte.config.js
export default {
  kit: {
    version: {
      name: Date.now(), // default to timestamp
      pollInterval: 60_000 // default to checking once a minute
    }
  }
};
```

File: packages/kit/src/core/build/build_client.js
Comment (pz-mxu) [reply packages/kit/src/core/build/build_client.js:32]: I thought that using VITE_APP_VERSION environment variable makes using a commit hash already very easy, but I agree that having all in one place in the config is also good. I propose that we set VITE_APP_VERSION to the value configured in the config file, but if there is nothing, it can also be set via environment. the fallback to `Date.now()`is later anyway.

File: packages/kit/src/runtime/client/renderer.js
Comment (pz-mxu) [reply packages/kit/src/runtime/client/renderer.js:74]: I added that catch because of error messages in the console in dev. Since dev is handled differently now, this is not necessary anymore.

**Comment by pz-mxu** [comment]
> Thank you. Thinking on this further, I'm reaching the conclusion that there's no reason for any behaviour other than automatic reloading. This is an area where the framework needs to be a little bit opinionated and do the right thing, since most developers don't really think about long session handling until they get bitten. Since client-side routing is best thought of as a progressive enhancement, falling back to a full page refresh should be perfectly acceptable in a well-written app.
> 
> To that end, I think we should get rid of the `onError` option (though if there's a compelling reason not to, I would be interested to hear it).

I totally agree, but I read the discussion on #87 and many people requested more configurability. For me personally, the reload is all I need and using SvelteKit in 4 projects already, that solution has helped with the problems after deploy. I don't know how the final decision on this is made and if I should remove the option right away or wait.

**Comment by pz-mxu** [comment]
> I didn't see any real arguments in favour of configurability in #87, just people working around the problem by making static assets outlive the deployment, which is a partial solution at best (since it ensures users will see stale content, and requires API compatibility between versions).
> 
> My inclination would be to _not_ make it configurable, and see if that causes problems. We can always add configuration in future, but removing unnecessary or problematic configuration is extremely hard.

Ok, I am nearly done with the changes - configurable version and poll interval. Will remove the configurability of reload which will make the changes much smaller, too.

File: packages/kit/src/core/build/build_client.js
Comment (Rich-Harris) [inline packages/kit/src/core/build/build_client.js:34]: Config is always fully defined — defaults should be specified in `options.js` (I realise this might make testing slightly more finicky)

```suggestion
	process.env.VITE_APP_VERSION = config.kit.version.name;
	process.env.VITE_APP_VERSION_FILE = `${config.kit.appDir}/version.json`;
	process.env.VITE_APP_VERSION_POLL_INTERVAL = `${config.kit.version.pollInterval}`;
```

**Comment by pz-mxu** [comment]
I don't know if passing the version, poll interval and version file's location as internal environment variables is acceptable, but I didn't come up with any better solution without adding more parameters passed to Renderer.

File: packages/kit/src/core/config/options.js
Comment (Rich-Harris) [inline packages/kit/src/core/config/options.js:259]:
```suggestion
				name: string(Date.now().toString()),
```

Comment (pz-mxu) [reply packages/kit/src/core/config/options.js:259]: But this will break the tests, because the timestamp will not be the same.

File: packages/kit/src/runtime/app/env.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/app/env.js:20]: it's guaranteed to exist and be a string

```suggestion
export const version = import.meta.env.VITE_APP_VERSION;
```

File: packages/kit/src/core/config/options.js
Comment (pz-mxu) [reply packages/kit/src/core/config/options.js:259]: I had it that way before.

File: packages/kit/src/runtime/app/env.js
Comment (Rich-Harris) [reply packages/kit/src/runtime/app/env.js:20]: ah, I see — it's not being set in dev. I think it's probably easier to set it to something dev-specific in dev, that way we have fewer points of divergence

File: packages/kit/src/core/config/options.js
Comment (Rich-Harris) [reply packages/kit/src/core/config/options.js:259]: it might require some finessing (overriding/deleting the value during testing, a la `remove_keys`, or monkey-patching `Date.now()` or something) but i think it's probably preferable to be able to make guarantees about the shape of the config

File: packages/kit/src/core/adapt/prerender/fixtures/large-page/input.html
Comment (Rich-Harris) [inline packages/kit/src/core/adapt/prerender/fixtures/large-page/input.html:1798]: we don't need to keep this file up to date. in fact i think we should remove it from the repo soon (it was originally there for benchmarking purposes)

```suggestion
```

File: packages/kit/src/core/adapt/prerender/fixtures/large-page/output.json
Comment (Rich-Harris) [inline packages/kit/src/core/adapt/prerender/fixtures/large-page/output.json:93]: ditto

```suggestion
```

**Review by pz-mxu — COMMENTED** [review]
This causes the type check to fail.

`export const version = import.meta.env.VITE_APP_VERSION;`

**Comment by Rich-Harris** [comment]
I don't think we need to expose `version` at all, to be honest. We could add it in future if it's useful but it would keep this PR simpler if we got rid of it

File: documentation/docs/14-configuration.md
Comment (Rich-Harris) [inline documentation/docs/14-configuration.md:73]: After discussion with other maintainers, this should probably default to `0`, which would mean 'no polling'

File: packages/kit/src/core/adapt/prerender/fixtures/large-page/input.html
Comment (Rich-Harris) [inline packages/kit/src/core/adapt/prerender/fixtures/large-page/input.html:1735]: removing this change (see #3661)

```suggestion
								returns <code>{ navigating, page, session }</code>. This needs to be called at the
```

File: packages/kit/src/runtime/client/renderer.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/renderer.js:153]: as long as we set `process.env.VITE_APP_VERSION_POLL_INTERVAL` in dev as well as build (albeit to 0, i.e. no polling) we can simplify this (note that it also changes `versionPollInterval` to `version_poll_interval` to be consistent with internal variable names elsewhere in the codebase):

```suggestion
		const version_poll_interval = +import.meta.env.VITE_APP_VERSION_POLL_INTERVAL;
```

File: packages/kit/src/runtime/client/renderer.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/renderer.js:168]: see previous

```suggestion
				version_poll_interval
```

File: packages/kit/src/runtime/client/renderer.js
Comment (pz-mxu) [reply packages/kit/src/runtime/client/renderer.js]: Gives type error now, too.

File: packages/kit/src/runtime/client/renderer.js
Comment (benmccann) [inline packages/kit/src/runtime/client/renderer.js:51]: perhaps this should only be done if `interval > 0`

Comment (pz-mxu) [reply packages/kit/src/runtime/client/renderer.js:51]: Already changed that while setting default interval to 0. Not yet pushed ;-)

File: documentation/docs/05-modules.md
Comment (Conduitry) [inline documentation/docs/05-modules.md:74]: This should be updated with the new default behavior.

`updated` should be added to the list of `import`s above from `'$app/stores'`.

For consistency with the other elements in this list, we should remove the newline before `- updated`.

Comment (Rich-Harris) [reply documentation/docs/05-modules.md:74]: I didn't actually clock that the store was writable. I don't think that's a good idea, I think it should be a readable store with a `check` method

Comment (Rich-Harris) [reply documentation/docs/05-modules.md]:
```suggestion
- `updated` is a [readable store](https://svelte.dev/tutorial/readable-stores) whose initial value is false. If [`version.pollInterval`](#configuration-version) is a non-zero value, SvelteKit will poll for new versions of the app and update the store value to `true` when it detects one. `updated.check()` will force an immediate check, regardless of polling.
```

File: documentation/docs/05-modules.md
Comment (Rich-Harris) [inline documentation/docs/05-modules.md:73]:
```suggestion
- `updated` is a [readable store](https://svelte.dev/tutorial/readable-stores) whose initial value is false. If [`version.pollInterval`](#configuration-version) is a non-zero value, SvelteKit will poll for new versions of the app and update the store value to `true` when it detects one. `updated.check()` will force an immediate check, regardless of polling.
```

File: documentation/docs/05-modules.md
Comment (pz-mxu) [reply documentation/docs/05-modules.md]: The only reason I made it writable is to enable people to add custom checks while still using that store. Since the store is used when routing errors occur, that behavior can be manually triggered.

File: packages/kit/src/runtime/app/stores.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/app/stores.js:89]: we should make this a readable store

```suggestion
```

File: packages/kit/src/runtime/app/stores.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/app/stores.js:96]:
```suggestion
```

File: packages/kit/src/runtime/client/renderer.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/renderer.js:59]:
```suggestion
```

File: packages/kit/src/runtime/server/page/render.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/server/page/render.js:91]:
```suggestion
		const updated = readable(false);
		updated.check = () => {};
```

Comment (Rich-Harris) [reply packages/kit/src/runtime/server/page/render.js:91]: we could in fact hoist this outside the function and reuse it, since its value will never change

File: packages/kit/types/ambient-modules.d.ts
Comment (Rich-Harris) [inline packages/kit/types/ambient-modules.d.ts:113]:
```suggestion
		updated: typeof updated;
```

File: packages/kit/types/ambient-modules.d.ts
Comment (Rich-Harris) [inline packages/kit/types/ambient-modules.d.ts:140]:
```suggestion
	export const updated: Readable<boolean> & { check: () => boolean };
```

File: documentation/docs/14-configuration.md
Comment (Rich-Harris) [inline documentation/docs/14-configuration.md:72]:
```suggestion
			name: Date.now().toString(),
```

File: documentation/docs/14-configuration.md
Comment (Rich-Harris) [inline documentation/docs/14-configuration.md:266]: Probably worth explaining what this config is for

```suggestion

Client-side navigation can be buggy if you deploy a new version of your app while people are using it. If the code for the new page is already loaded, it may have stale content; if it isn't, the app's route manifest may point to a JavaScript file that no longer exists. SvelteKit solves this problem by falling back to traditional full-page navigation if it detects that a new version has been deployed, using the `name` specified here (which defaults to a timestamp of the build).

If you set `pollInterval` to a non-zero value, SvelteKit will poll for new versions in the background and set the value of the [`updated`](#modules-$app-stores) store to `true` when it detects one.
```

**Comment by Rich-Harris** [comment]
Made a small tweak so that the interval is always set (albeit to 0) even in dev. 

One last (I think) thing — we probably need to use the `VITE_SVELTEKIT_` namespace for consistency with existing code.

**Comment by pz-mxu** [comment]
> Made a small tweak so that the interval is always set (albeit to 0) even in dev.

Great, I already looked for the place where this has to go.

File: documentation/docs/14-configuration.md
Comment (Rich-Harris) [inline documentation/docs/14-configuration.md:269]:
```suggestion
If you set `pollInterval` to a non-zero value, SvelteKit will poll for new versions in the background and set the value of the [`updated`](#modules-$app-stores) store to `true` when it detects one.

```

**Review by Rich-Harris — APPROVED** [review]

**Comment by Rich-Harris** [comment]
We did it! Thanks so much for this mammoth effort @pz-mxu — this is a great addition to the framework

**Comment by pz-mxu** [comment]
Happy I could be of help.

**Comment by jycouet** [comment]
I'm not sure it's the right way to comment here, but I'll give it a try.

I'm precisely looking at `kit.browser.router` `onError` `reload` (Because I'm deploying a lot atm 🥳)
But on `SvelteKit v1.0.0-next.295` (latest now) I can't figure out how to have this excellent reloading behavior.

Could you help me?

**Comment by Antonio-Bennett** [comment]
Hey @jycouet After reading through this seems like it is a configuration on `svelte.config.js` specifically https://kit.svelte.dev/docs/configuration#version

**Comment by jycouet** [comment]
I did change `pollInterval` to 5000 to see.
But it's not doing `onError` => `reload`. Actually, as I'm not using it (`updated`), it's not doing anything 😅

Now, using `updated` I'm deploying to see if it's better. 
In `__layout.svelte`
```ts
import { updated } from '$app/stores';

updated.subscribe(async (value) => {
  if (value) {
    // Hard refresh when version is not matching. 
    // I can test only on prod! (dev or preview, this is not working) https://github.com/sveltejs/kit/issues/3957
    // Todo:
    // Nicer solution here one day: https://github.com/sveltejs/kit/issues/3666
    // or move to web workers? https://github.com/sveltejs/kit/issues/3667
    window.location.reload();
  }
});
```

=> It didn't refresh my page on the next deployment! I didn't see an error as well... 🤔

It's really hard to debug as I don't know how to reproduce locally.
Do you have ideas?

**Comment by frederikhors** [comment]
> It didn't refresh my page on the next deployment!

Still not working?

**Comment by jycouet** [comment]
Seems like it's not working :(

**Comment by frederikhors** [comment]
I'm on 1.0.0-next.295 and `updated` store works. I tried with 5000 ms too...

**Comment by jycouet** [comment]
🥳🥳🥳 I found out!!!

You need: `NODE_ENV: production` _(I had an issue in my CI that this fixed now 👌)_.
Another excellent info, the **version** can be found in: `VITE_SVELTEKIT_APP_VERSION`

**Comment by EskelCz** [comment]
@Rich-Harris Isn't there an issue of the app losing its state? So for example a long text the user has written. I would rather show a warning message that the app was meanwhile updated and ask the user to refresh when it's convenient to them.

Also I love that SvelteKit at least has a solution, but when using Svelte+Vite only, there's no solution to be found. :(
This is the issue thread there: https://github.com/vitejs/vite/issues/11804

Thanks for your consideration.

**Pull request merged**
